packages/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx

```diff
@@ -76,11 +94,29 @@ export const useGridVirtualScroller = () => {
 
   useResizeObserver(mainRef, () => apiRef.current.resize());
 
+  /*
+   * Scroll context logic
+   * ====================
+   * We only render the cells contained in the `renderContext`. However, when the user starts scrolling the grid
+   * in a direction, we want to render as many cells as possible in that direction, as to avoid presenting white
+   * areas if the user scrolls too fast/far and the viewport ends up in a region we haven't rendered yet. To render
+   * more cells, we store some offsets to add to the viewport in `scrollCache.buffer`. Those offsets make the render
+   * context wider in the direction the user is going, but also makes the buffer around the viewport `0` for the
+   * dimension (horizontal or vertical) in which the user is not scrolling. So if the normal viewport is 8 columns
+   * wide, with a 1 column buffer (10 columns total), then we want it to be exactly 8 columns wide during vertical
+   * scroll.
+   * However, we don't want the rows in the old context to re-render from e.g. 10 columns to 8 columns, because that's
+   * work that's not necessary. Thus we store the context at the start of the scroll in `frozenContext`, and the rows
+   * that are part of this old context will keep their same render context as to avoid re-rendering.
+   */
```

Comment:
![image](https://github.com/mui/mui-x/assets/1423607/0f75d837-59b2-4ba3-9265-8e820b62635a)

Reply:
Thanks for the detailed comment and visualization!

Reply:
Sure, I couldn't find names good enough to explain the logic alone. And I made the graph before implementing the thing, helps my brain understand which objects/concepts I'll need. Also I just like graphs.

Reply:
Yes, pretty useful to understand what's going on 👍

Reply:
> but also makes the buffer around the viewport `0` for the dimension (horizontal or vertical) in which the user is not scrolling

Nice. Would this work on iOS too? It might cause issues due to the [diagonal scrolling](https://github.com/mui/mui-x/issues/11230), but it will probably be solved automatically when we solve 11230, right?

Reply:
Yes, we'll need to solve diagonal scrolling, but that shouldn't be too hard.